src/alvin0319/OffHand/OffHand.php

```diff
@@ -30,84 +30,87 @@
 	}
 
+	/**
+	 * @ignoreCancelled true
+	 * @priority MONITOR
+	 */
 	public function onDataPacketReceive(DataPacketReceiveEvent $event) : void{
 		$packet = $event->getPacket();
 		$player = $event->getPlayer();
-		if($packet instanceof MobEquipmentPacket){
-			if($packet->windowId === ContainerIds::OFFHAND){
-				$inv = $this->getOffHandInventory($player);
-				if($inv instanceof OffHandInventory){
-					$inv->setItemInOffHand($packet->item);
-					$event->setCancelled();
-				}
-			}
-		}
+        	if($packet instanceof MobEquipmentPacket and $packet->windowId == ContainerIds::OFFHAND){
+            		$offhand = $this->getOffhandInventory($player);
+            		if(!$offhand->getItem(0)->equalsExact($packet->item)){
+                		$offhand->sendContents($player);
+                		$event->setCancelled();
+                		return;
+            		}
+            		$offhand->setItem(0, $packet->item);
+        	}
 	}
-
+	/**
+	 * @ignoreCancelled true
+	 * @priority MONITOR
+	 */
 	public function onDataPacketSend(DataPacketSendEvent $event) : void{
-		$packet = $event->getPacket();
-		if($packet instanceof AddPlayerPacket){
-			if(($player = $this->getServer()->getPlayerExact($packet->username)) !== null){
-				$this->getOffHandInventory($event->getPlayer())->sendMobEquipmentPacket($player);
-				$this->getOffHandInventory($event->getPlayer())->sendMobEquipmentPacket($event->getPlayer());
-			}
-		}
+        	$packet = $event->getPacket();
+        	$player = $event->getPlayer();
+        	if($packet instanceof AddPlayerPacket){
+            		$this->getOffhandInventory($player->getServer()->getPlayerExact($packet->username) ?? $player)->sendItem([$player]);
+        	}
 	}
-
+	/**
+	 * @priority MONITOR
+	 */
 	public function onDeath(PlayerDeathEvent $event) : void{
-		$player = $event->getPlayer();
-		$drops = $event->getDrops();
-		if(!$event->getKeepInventory()){
-			$drops = array_merge($drops, $this->getOffHandInventory($player)->getContents(false));
-			$event->setDrops($drops);
-			$this->getOffHandInventory($player)->clearAll();
+        	$player = $event->getPlayer();
+        	$offhand = $this->getOffhandInventory($player);
+        	if(!$event->getKeepInventory() and !empty($event->getDrops())){
+            		$event->setDrops(array_merge($event->getDrops(), array_filter([$offhand->getItem(0)], function(Item $item): bool{
+                		return !$item->hasEnchantment(Enchantment::VANISHING);
+            		})));
+            		$offhand->clearAll();
 		}
+        }
+	private static $offhand = [];
+	public function getOffHandInventory(Human $player): OffHandInventory{
+        	$UUID = $player->getUniqueId()->toString();
+        	$inventory = self::$offhand[$UUID] = self::$offhand[$UUID] ?? new OffhandInventory($player);
+        	if($player instanceof Player){
+            		$player->addWindow($inventory, ContainerIds::OFFHAND, true);
+        	}
+        	if($player->namedtag->hasTag("Offhand", ListTag::class)){
```

Comment:
and why use ListTag? This will cause BC break.

Reply:
In Vanilla Minecraft it uses ListTag instead of a CompoundTag, so I added that detail to be vanilla-alike

Reply:
Please use CompoundTag for Backwards compatibilty

Reply:
Oh true